pandas/core/arrays/period.py

```diff
@@ -83,88 +103,216 @@ def wrapper(self, other):
+        # coerce freq to freq object, otherwise it can be coerced elementwise
+        # which is slow
+        if freq:
+            freq = Period._maybe_convert_freq(freq)
 
-        return cls._simple_new(values, freq, **kwargs)
+        if data is None:
+            if ordinal is not None:
+                data = np.asarray(ordinal, dtype=np.int64)
+            else:
+                data, freq = cls._generate_range(start, end, periods,
+                                                 freq, fields)
+            return cls._from_ordinals(data, freq=freq)
+
+        if isinstance(data, (cls, PeriodIndex)):
+            if freq is None or freq == data.freq:  # no freq change
+                freq = data.freq
+                data = data._ndarray_values
+            else:
+                base1, _ = _gfc(data.freq)
+                base2, _ = _gfc(freq)
+                data = libperiod.period_asfreq_arr(data._ndarray_values,
+                                                   base1, base2, 1)
+            if copy:
+                data = data.copy()
+            return cls._simple_new(data, freq=freq)
+
+        # not array / index
+        if not isinstance(data, (np.ndarray, PeriodIndex,
+                                 DatetimeIndex, Int64Index)):
+            if is_scalar(data):
+                raise TypeError('{0}(...) must be called with a '
+                                'collection of some '
+                                'kind, {1} was passed'.format(cls.__name__,
+                                                              repr(data)))
+
+            # other iterable of some kind
+            if not isinstance(data, (list, tuple)):
+                data = list(data)
+
+            data = np.asarray(data)
+
+        # datetime other than period
+        if is_datetime64_dtype(data.dtype):
+            data = dt64arr_to_periodarr(data, freq, tz)
+            return cls._from_ordinals(data, freq=freq)
+
+        # check not floats
+        if lib.infer_dtype(data) == 'floating' and len(data) > 0:
+            raise TypeError("PeriodIndex does not allow "
+                            "floating point in construction")
+
+        # anything else, likely an array of strings or periods
+        data = ensure_object(data)
+        return cls._from_periods(data, freq=freq)
 
     @classmethod
     def _simple_new(cls, values, freq=None, **kwargs):
         """
         Values can be any type that can be coerced to Periods.
         Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
         """
-
         if not is_integer_dtype(values):
             values = np.array(values, copy=False)
             if len(values) > 0 and is_float_dtype(values):
                 raise TypeError("{cls} can't take floats"
                                 .format(cls=cls.__name__))
             return cls(values, freq=freq)
 
-        return cls._from_ordinals(values, freq)
+        return cls(values, freq=freq)
```

Comment:
Does calling the base constructor for simple_new make sense?

Reply:
I think I commented it somewhere else, but I think we should look into just merging the two in the default `__init__`. `_simple_new` is basically doing exactly the same